### PR TITLE
feat(PF4-Dropdown): Dropdown keyboard interaction

### LINKED ISSUE #557: Dropdown keyboard interaction
The current implementation of the PF4 Dropdown component needs some enhancements to the keyboard interaction. The following suggestions were noted in the [original PR for Dropdown](https://github.com/patternfly/patternfly-react/pull/506):

- When the menu is visible and the user hits `Esc`
  - Current Behavior: Nothing happens
  - Expected Behavior: menu is hidden and focus is placed on the toggle
- **_updated_** When the menu is visible, the last menu item has focus, and the user hits `Tab`
  - Current Behavior: The next element that can receive focus has focus, and the menu remains visible.
    - This can present usability issues for keyboard-only users who are sighted, since they might not be able to see what has focus if the menu is obscuring it. But we also don't want to automatically close the menu without explicit action, which could also be an issue.
  - ~~Expected Behavior: The next element that can receive focus has focus, and the menu is hidden. Then if the user hits `Shift` + `Tab`, the menu becomes visible again, and the last menu item receives focus.~~
  - Expected Behavior: The first item in the menu that can receive focus receives focus
    - Likewise, when the user is using `Shift` + `Tab` to move up through the menu, when focus leaves the first menu item, it should be placed on the last menu item.
- **_new_** When the toggle has focus and the user hits `Enter`/`Space`:
  - Expected Behavior: The menu displays and the first item in the menu that can receive focus receives focus

**Please note:** There are different variations of a "Dropdown" component. Similar to [Bootstrap](https://getbootstrap.com/docs/4.1/components/dropdowns/), the [PF4 dropdown](https://pf-next.com/components/Dropdown/examples/) is implemented so that anything can be included in the popup section. This means that the elements that receive focus might not be links or buttons, but could be other things like form elements. 

[Additional comments were posted](https://github.com/patternfly/patternfly-react/pull/506#issuecomment-416208197) in response to this suggestion. Any feedback or alternate suggestions are welcome. Our goal is to provide a solution that can accommodate all users, ranging from mouse users, sighted keyboard-only users, and screen reader users.'

### PR DESCRIPTION
**Pull request by michaelkro**
Fixes #557

**What**:
1. Scope `esc` keypress to the dropdown containing the currently focused DOM element
2. Automatically close dropdown if user tabs all the way through the menu (forwards or backwards)
3. Automatically reopen dropdown and focus last menu item under the following conditions.  
    - Keyboard user tabs all the way through the menu (so that it autocloses)
    - Keyboard user uses `shift + tab` to move focus back to the recently closed dropdown

**TODO**:

1.  ~Add type definitions~
2.  ~Add tests~

### CODE REVIEW

**Comment by jgiardino** [comment]
Thanks for working on this! The behavior works as I expect it to. 

What I'm curious about is whether the way we're implementing keyboard interaction here is something that works for other components. 

And also whether we can put a dropdown inside a modal and have the Escape key work as expected, but this might require a separate PR to figure this out.

**Comment by jeff-phillips-18** [comment]
Can we specify clearly what about the dropdown didn't work correctly before? Looking at #557 and the current Dropdown, I see most of this working prior to this PR.

I see a lot of state control being added and wondering how much is really necessary.

**Comment by michaelkro** [comment]
@jgiardino, I can easily remove those attributes, but not sure if it would be best to do that in a separate PR.  I'll let the other reviewers make the call.

@jeff-phillips-18, so this is interesting.  It seems that something about the live examples is affecting interactions with our components.  If you start up your local dev environment with

```sh
GATSBY_LIVE_EXAMPLES=false yarn start:pf4
```

You'll see the behavior #557 mentions, specifically

1.  The dropdown does not automatically collapse when tabbing away
2.  Hitting the esc key will focus the last dropdown example (the kebab example) instead of returning focus to the dropdown that was closed

The "auto-open" on `shift + tab` is a feature that was added to satisfy the requirements of #557

**Comment by tlabaj** [comment]
@jgiardino @michaelkro I ma currently working on  #718

File: packages/patternfly-4/react-core/src/components/Dropdown/Dropdown.js [hunk 1]
```diff
@@ -63,16 +100,33 @@ class Dropdown extends React.Component {
           isPlain && styles.modifiers.plain,
           direction === DropdownDirection.up && styles.modifiers.top,
           position === DropdownPosition.right && styles.modifiers.alignRight,
-          isOpen && styles.modifiers.expanded,
+          isOpen && focusedItemRef && styles.modifiers.expanded,
           className
         )}
         ref={ref => {
           this.parentRef = ref;
```
Comment (dmiller9911) [inline packages/patternfly-4/react-core/src/components/Dropdown/Dropdown.js]: refs should be using the new syntax `React.createRef()` and then they are accessed through `this.parentRef.current` https://reactjs.org/docs/refs-and-the-dom.html#creating-refs we are only supporting the latest react right now so it should be safe to use.

**Comment by jeff-phillips-18** [comment]
@michaelkro So what caused the difference between http://patternfly-react.surge.sh/patternfly-4/components/dropdown and running current locally?

**Comment by jschuler** [comment]
@michaelkro If react-live is causing an issue with your examples, it is also possible to turn it off per example by adding `static live = false;` in the example file. In case this needs to move forward while looking at the react-live interaction issue separately.

**Comment by dgutride** [comment]
Please do not merge this - Jeff and I would like to review the design requirements and implementation with @jgiardino  and @michaelkro before this goes in.

**Comment by michaelkro** [comment]
Closing as we are taking a different approach

**Pull request closed without merging**